bluetooth: host: Run-time interface for setting RPA timeout

mtpr-ot opened the pull request:
Added public interface for setting host controlled RPA timeout at
run-time.

Setting the RPA timeout allows control of how often the host changes
the resolvable private address to reduce trackability. This is used
by non-directed RPA advertising, with no controller privacy bonding
information. Default timeout is defined by CONFIG_BT_RPA_TIMEOUT.

Signed-off-by: Morten Priess <mtpr@oticon.com>

jhedberg commented:
Could you fix the subject prefix? This is not touching any controller code.

mtpr-ot commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -6692,3 +6713,20 @@ int bt_le_oob_get_sc_data(struct bt_conn *conn,
+
+	return err;
+}
+#endif /* CONFIG_BT_PRIVACY */
```
Regarding bt_set_rpa_timeout: I could move this to vendor specific interface, if it doesn't fit in...

jhedberg commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -209,3 +209,5 @@ int bt_le_adv_start_internal(const struct bt_le_adv_param *param,
 			     const bt_addr_le_t *peer);
 void bt_le_adv_resume(void);
+
+int bt_set_rpa_timeout(u16_t timeout);
```
Do you intend this to be a public or internal API? If it is to be public then it needs to go to `include/bluetooth/` since `hci_core.h` is an internal header file. Where do you expect this to be called.

In general your commit message could use a lot more "flesh on the bones", in particular answering the most important question any commit message should answer, i.e. "why?" (the answer to "what?" can be found simply by looking at the diff, so it's less valuable, but still ok to have in the commit message).

mtpr-ot replied:
> Do you intend this to be a public or internal API? 

I intended it to be a public interface because I think it would be a useful addition. however, as this is a specific need for our product, I could easily accept making it vendor specific. Moved prototype to bluetooth.h.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -6692,3 +6713,20 @@ int bt_le_oob_get_sc_data(struct bt_conn *conn,
+{
+	int err;
+
+	/* Restart timer even with new value */
```
This is a restart, but no action taken? Then it would be remaning_value + new_value until RPA timeout happens again. I think it should be refreshed immediately here, and then again in new_value in the future.

If we refresh RPA here we need to handle conn create using the current RPA, as the random address cannot be updated while the controller initiatior is active.

mtpr-ot replied:
> This is a restart, but no action taken? Then it would be remaning_value + new_value until RPA timeout happens again. I think it should be refreshed immediately here, and then again in new_value in the future.
> 
> If we refresh RPA here we need to handle conn create using the current RPA, as the random address cannot be updated while the controller initiatior is active.

Do you mean "elapsed + new"? Then yes, that would potentially extend the RPA refresh. It would be a good idea to refresh the RPA immediately or by using a short "inital work delay" to have the regular timeout do it and restart with new value.
I'm not sure I can see how to prevent RPA refresh while controller initiatior is active in all cases?

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -6692,3 +6713,20 @@ int bt_le_oob_get_sc_data(struct bt_conn *conn,
 #endif
+
+#if defined(CONFIG_BT_PRIVACY)
+int bt_set_rpa_timeout(u16_t timeout)
```
Related to some other work (which is incomplete at the  moment) I think we should restrict the timeout to always be larger than CONFIG_BT_CONN_CREATE_TIMEOUT.

For refrence incomplete work is here: https://github.com/zephyrproject-rtos/zephyr/pull/21714

mtpr-ot replied:
OK

joerchan reviewed (changes requested):
Need to consider some already existing problems with the RPA timeout in the host

mtpr-ot committed:
bluetooth: host: Run-time interface for setting RPA timeout

Added public interface for setting host controlled RPA timeout at
run-time.

Setting the RPA timeout allows control of how often the host changes
the resolvable private address to reduce trackability. This is used
by non-directed RPA advertising, with no controller privacy bonding
information. Default timeout is defined by CONFIG_BT_RPA_TIMEOUT.

Signed-off-by: Morten Priess <mtpr@oticon.com>

joerchan commented:
@mtpr-ot I think it would be helpful to understand why the RPA timeout value needs to be changed dynamically at runtime. Why can't it be just a Kconfig option?

When it comes to the complexity of this function. The requirement of the use-case becomes important. The easiest would be to just let the old timeout value expire, and the use the new value from RPA timeout. Note that this is currently also the behaviour of the controller as far as I can see.
If you have to cancel the current time, refresh RPA and restart with new value we need to be more careful. 

If what you need is simply to restart the RPA timeout and not use a different value, then `bt_le_oob_get_local` could be used. But that can only be called when all all roles using the random address are inactive (advertiser/scanner/initiator/).

mtpr-ot commented:
I just had a meeting with the stakeholders for this requirement. We agreed that it will be sufficient for now to change the RPA timeout via Kconfig. We will modify our requirments instead, and abandon this PR for now.
@joerchan, @jhedberg, thanks for your valid input, and sorry for taking your time with this PR.
